synapse/handlers/cas_handler.py

```diff
@@ -236,24 +258,54 @@ async def _map_cas_user_to_matrix_user(
 
         Args:
             remote_user_id: The username from the CAS response.
-            display_name: The display name from the CAS response.
+            attributes: Additional attributes from the CAS response.
             user_agent: The user agent of the client making the request.
             ip_address: The IP address of the client making the request.
 
+        Raises:
+            MappingException: if there was an error while mapping some properties
+
         Returns:
              The user ID associated with this response.
         """
 
+        # Note that CAS does not support a mapping provider, so the logic is hard-coded.
         localpart = map_username_to_mxid_localpart(remote_user_id)
-        user_id = UserID(localpart, self._hostname).to_string()
-        registered_user_id = await self._auth_handler.check_user_exists(user_id)
-
-        # If the user does not exist, register it.
-        if not registered_user_id:
-            registered_user_id = await self._registration_handler.register_user(
-                localpart=localpart,
-                default_display_name=display_name,
-                user_agent_ips=[(user_agent, ip_address)],
+        display_name = attributes.pop(self._cas_displayname_attribute, None)
+
+        async def cas_response_to_user_attributes(failures: int) -> UserAttributes:
+            """
+            Map from CAS attributes to user attributes.
+            """
+            # Due to the grandfathering logic matching any previously registered
+            # mxids it isn't expected for there to be any failures.
+            if failures:
+                raise RuntimeError("CAS is not expected to de-duplicate Matrix IDs")
```

Comment:
This code should be hit since in the case of a conflict causing a failure we should have already have grandfathered in `grandfather_existing_users`.